src/range/bytes.rs

```diff
@@ -0,0 +1,542 @@
+use crate::{Error, StatusCode};
+
+use std::fmt::{self, Debug, Display};
+use std::str::FromStr;
+
+pub(crate) const ACCEPT_RANGE_VALUE: &str = "bytes";
+pub(crate) const RANGE_PREFIX: &str = "bytes=";
+pub(crate) const CONTENT_RANGE_PREFIX: &str = "bytes ";
+
+/// The representation of a single HTTP byte range.
+///
+/// # Specifications
+///
+/// - [RFC 7233, section 2.1: Range](https://tools.ietf.org/html/rfc7233#section-2.1)
+/// - [RFC 7233, Appendix D: Collected ABNF](https://tools.ietf.org/html/rfc7233#appendix-D)
+#[derive(Default, Debug, Clone, Eq, PartialEq, Copy)]
+pub struct BytesRange {
+    /// The range start.
+    ///
+    /// If empty the ends indicates a relative start
+    /// from the end of the document.
+    pub start: Option<u64>,
+    /// The range end.
+    ///
+    /// If empty the range goes through the end
+    /// of the document.
+    pub end: Option<u64>,
+}
+
+impl BytesRange {
+    /// Create a new instance with start and end.
+    pub fn new<S, E>(start: S, end: E) -> Self
+    where
+        S: Into<Option<u64>>,
+        E: Into<Option<u64>>,
+    {
+        Self {
+            start: start.into(),
+            end: end.into(),
+        }
+    }
+
+    /// Returns true if the range's bounds match the given document size.
+    pub fn match_size(&self, size: u64) -> bool {
+        if let Some(start) = self.start {
+            if start > size - 1 {
+                return false;
+            }
+        }
+        if let Some(end) = self.end {
+            if end > size - 1 {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+impl Display for BytesRange {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}-{}",
+            self.start
+                .map(|v| v.to_string())
+                .unwrap_or_else(String::new),
+            self.end.map(|v| v.to_string()).unwrap_or_else(String::new),
+        )
+    }
+}
+
+impl FromStr for BytesRange {
```

Comment:
Maybe I should just provide a `pub(crate) fn from_str` implementation and not do it as `impl FromStr` to remain consistant with the rest of the code.